service/history/ndc/history_replicator_test.go

```diff
@@ -1007,3 +1007,855 @@ func Test_applyEvents_defaultCase_errorAndDefault(t *testing.T) {
+		},
+		"Case4: error when calling stateBuilder.ApplyEvents": {
+			mockDomainCacheAffordance: func(mockDomainCache *cache.MockDomainCache) {
+				mockDomainCache.EXPECT().GetDomainByID(gomock.Any()).
+					Return(cache.NewGlobalDomainCacheEntryForTest(nil, nil, nil, 1), nil).Times(1)
+			},
+			mockStateBuilderAffordance: func(mockStateBuilder *execution.MockStateBuilder) {
+				mockStateBuilder.EXPECT().ApplyEvents(
+					gomock.Any(),
+					gomock.Any(),
+					gomock.Any(),
+					gomock.Any(),
+					gomock.Any(),
+				).Return(nil, fmt.Errorf("test error")).Times(1)
+			},
+			mockReplicationTaskAffordance: func(mockReplicationTask *MockreplicationTask) {
+				mockReplicationTask.EXPECT().getLogger().Return(log.NewNoop()).Times(3)
+				mockReplicationTask.EXPECT().getDomainID().Return("test-domain-id").Times(2)
+				mockReplicationTask.EXPECT().getExecution().Return(&types.WorkflowExecution{
+					WorkflowID: "test-workflow-id",
+					RunID:      "test-run-id",
+				}).Times(1)
+				mockReplicationTask.EXPECT().getEvents().Return(nil).Times(1)
+				mockReplicationTask.EXPECT().getNewEvents().Return(nil).Times(1)
+			},
+			mockTransactionManagerAffordance: func(mockTransactionManager *MocktransactionManager) {},
+			mockShardContextAffordance:       func(mockShard *shard.MockContext) {},
+			expectError:                      fmt.Errorf("test error"),
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+
+			// Mock objects
+			mockDomainCache := cache.NewMockDomainCache(ctrl)
+			mockExecutionContext := execution.NewMockContext(ctrl)
+			mockMutableState := execution.NewMockMutableState(ctrl)
+			mockStateBuilder := execution.NewMockStateBuilder(ctrl)
+			mockReplicationTask := NewMockreplicationTask(ctrl)
+			mockTransactionManager := NewMocktransactionManager(ctrl)
+			mockShard := shard.NewMockContext(ctrl)
+			logger := log.NewNoop()
+
+			// Mock affordances
+			test.mockDomainCacheAffordance(mockDomainCache)
+			test.mockStateBuilderAffordance(mockStateBuilder)
+			test.mockReplicationTaskAffordance(mockReplicationTask)
+			test.mockTransactionManagerAffordance(mockTransactionManager)
+			test.mockShardContextAffordance(mockShard)
+
+			// Mock functions
+			mockNewMutableStateFn := func(
+				domainEntry *cache.DomainCacheEntry,
+				logger log.Logger,
+			) execution.MutableState {
+				return mockMutableState
+			}
+
+			mockNewStateBuilderFn := func(
+				state execution.MutableState,
+				logger log.Logger,
+			) execution.StateBuilder {
+				return mockStateBuilder
+			}
+
+			// Call the function under test
+			err := applyStartEvents(ctx.Background(), mockExecutionContext, func(err error) {}, mockReplicationTask,
+				mockDomainCache, mockNewMutableStateFn, mockNewStateBuilderFn, mockTransactionManager, logger, mockShard, cluster.Metadata{})
+
+			// Assertions
+			assert.Equal(t, test.expectError, err)
```

Comment:
There was an error when I changed it to Errorls, and the explaination is below (from GPT):

assert.ErrorIs is used to check for errors in the error chain, but this function relies on Go's error chaining mechanism (i.e., errors.Is). This means that the original error must be wrapped using errors.Wrap or fmt.Errorf("...: %w", err). If there is no such error chain, assert.ErrorIs will not be able to match.

In your code, the error is returned directly using fmt.Errorf("test error"), and in this case, assert.ErrorIs cannot match it because the error chaining mechanism was not used. If you do not intend to use error chaining, you can use assert.EqualError to compare the specific error message.